source/rpt/rpt_cell_reconstruction.cc

```diff
@@ -0,0 +1,501 @@
+      find_vertices_positions(level, cells_indexes);
+      calculate_counts();
+
+      // Find cell candidate for the position
+      // (reconstruction counts are in the counts range of the cells for every
+      // detector)
+      new_cells_indexes =
+        find_cells(level, particle_reconstruction_counts, cells_indexes);
+
+      // If no refined cells are candidates at current level, research has to
+      // stop and need to go back to parent level
+      if (new_cells_indexes[0] == cells_indexes[0])
+        {
+          still_new_candidates = false;
+          level--;
+        }
+
+      cells_indexes = new_cells_indexes;
+    }
+
+  // If many cells are candidate at the more refine level, the least squared
+  // method is used to get the best cell
+  int best_cell;
+  if (cells_indexes.size() > 1)
+    {
+      best_cell =
+        find_best_cell(level, particle_reconstruction_counts, cells_indexes);
+    }
+  else
+    best_cell = cells_indexes[0];
+
+  // Find the volume and the center position of the best cell
+  for (const auto &cell : triangulation.cell_iterators_on_level(level))
+    {
+      if (cell->index() == best_cell)
+        {
+          volume                  = cell->measure();
+          reconstruction_position = cell->center(true);
+        }
+    }
+
+  // Store reconstruction positions and volume prior exportation
+  reconstruction_positions.push_back(reconstruction_position);
+  cells_volumes.push_back(volume);
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::create_grid()
+{
+  TimerOutput::Scope t(computing_timer, "grid_creation");
+
+  // Generate cylinder (needs rotation and shift to get origin at the bottom
+  // with z towards top)
+  GridGenerator::subdivided_cylinder(triangulation,
+                                     5,
+                                     parameters.rpt_param.reactor_radius,
+                                     parameters.rpt_param.reactor_height / 2.);
+  GridTools::rotate(M_PI_2, 1, triangulation);
+  const Tensor<1, dim> shift_vector(
+    {0, 0, parameters.rpt_param.reactor_height / 2.});
+  GridTools::shift(shift_vector, triangulation);
+
+  // Add cylindrical manifold
+  const Tensor<1, dim>                direction({0, 0, 1});
+  const CylindricalManifold<dim, dim> manifold(direction, {0, 0, 1});
+  triangulation.set_manifold(0, manifold);
+
+  // Refine all cells
+  triangulation.prepare_coarsening_and_refinement();
+  triangulation.refine_global(
+    parameters.reconstruction_param.reactor_refinement);
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::find_vertices_positions(
+  unsigned int     level,
+  std::vector<int> parent_cell_indexes /* {-1} */)
+{
+  TimerOutput::Scope t(computing_timer, "finding_vertices_positions");
+
+  // Find the positions associated to the vertices of cell
+  for (const auto &cell : triangulation.cell_iterators_on_level(level))
+    {
+      bool parent_is_candidate = false;
```

Comment:
The only case that parent_is_candidate=false happens is when the algorithm wants to find the vertices of the coarse mesh at the beginning?

Reply:
Yes, or the list of candidate cells is empty (fail to find a cell at a higher refined level)
![image](https://user-images.githubusercontent.com/71884806/130085604-ad5a43f4-3cf6-42d6-abf1-f85803ccb550.png)

Reply:
Will add explaination